### PR TITLE
Feature/sling 9550 all scalars

### PR DESCRIPTION
**Pull request by etugarev**

### CODE REVIEW

File: src/main/java/org/apache/sling/graphql/core/scalars/SlingScalarsProvider.java [hunk 1]
```diff
@@ -46,55 +49,17 @@
...
+    @Reference
+    private SlingScalarConvertersProvider slingScalarConvertersProvider;
 
-    public Iterable<GraphQLScalarType> getCustomScalars(Map<String,ScalarTypeDefinition> schemaScalars) {
-        // Using just the names for now, not sure why we'd need the ScalarTypeDefinitions
-        return schemaScalars.keySet().stream()
-            .map(this::getScalar)
-            .filter(it -> it != null)
-            .collect(Collectors.toList());
+    public List<GraphQLScalarType> getCustomScalars() {
```
Comment (etugarev) [inline src/main/java/org/apache/sling/graphql/core/scalars/SlingScalarsProvider.java:55]: Injecting all at once

File: src/test/java/org/apache/sling/graphql/core/mocks/URLScalarConverter.java [hunk 2]
```diff
@@ -41,4 +41,14 @@
         final String testPrefix = getClass().getSimpleName() + " says:";
         return testPrefix + (value == null ? null : value.toExternalForm());
     }
+
+    @Override
+    public String getName() {
```
Comment (etugarev) [inline src/test/java/org/apache/sling/graphql/core/mocks/URLScalarConverter.java:46]: we have name and description in `SlingScalarConverter` here it's not yet implemented

**Comment by bdelacretaz** [comment]
We used to have a similar design for the SlingDataFetcher initially, with a single OSGi provider service which provided multiple data fetchers. Changing that to a design where SlingDataFetchers are individual OSGi services allowed me to remove a lot of code, especially in the [samples website](https://github.com/apache/sling-samples/tree/master/org.apache.sling.graphql.samples.website)

This makes me think that my original design (at commit 99062b5ba) where each SlingScalarConverter is an OSGi service, using the OSGi framework to select them, is more consistent with the data fetchers part, and also likely to lead to simpler client code.

What advantages do you see in your alternative design?

**Comment by stefangrimm** [comment]
@bdelacretaz We need something like the suggested alternative to decouple what we're implementing on AEM's side from the Sling Core implementation. It's an important design goal for us to be able to use our implementation with other GraphQL frameworks as well, without requiring framework-specific code adjustments.

With the current implementation, a scalar implementation would need to implement `SlingScalarConverter` directly, therefore adding a dependency on Sling.

The proposed solution would allow us to use framework-agnostic implementations of scalars for our stuff. We would then adapt these scalar implementations to something that's suitable for the respective framework within the adapter bundle (that is required for each GraphQL framework anyway).

There might be some other solutions (although I'm by far no OSGi expert, I guess we could register multiple instance of the same `SlingScalarConverter`-implementing wrapper component (one for each scalar) manually within the adapter, for example, but all the additional lifecycle-handling (when to register/unregister) frightens me). But from our perspective the proposed solution would be the simplest.

If the proposal is not acceptable for you, we can also reduce the support for scalars in our code to the absolute minimum. At least for now we could live with this.

Note that we don't have similar problems with data fetchers, as we're actually only using one single data fetcher and distribute it internally with the help of parameters (as part of the "fetch" directive) to our own, Sling-unaware data fetchers.

cc @etugarev

**Comment by stefangrimm** [comment]
Also, in addition to my previous comment:

IMHO scalars should not be defined globally, but per endpoint. Typically, they are part of the schema, and two schemas might use different implementations of a scalar with the same name. A typical user-defined scalar could be something around `Date` - with subtle differences for different schemas.

While AFAIU the currently proposed change doesn't take that into account, it could be easily extended to pass the endpoint resource to consider for creating the list of scalars.

**Comment by bdelacretaz** [comment]
> I guess we could register multiple instance of the same SlingScalarConverter-implementing wrapper component (one for each scalar) manually within the adapter, for example, but all the additional lifecycle-handling (when to register/unregister) frightens me). But from our perspective the proposed solution would be the simplest.

In the end I think this is the same as the provider service that you are suggesting: you end up with a wrapper for each scalar converter that adapts it to the Sling API.

The lifecycle handling is not complicated if you need to wrap "arbitrary scalar converters": just register an OSGi `SlingScalarProvider` service for each and you're good. This can be done either with Declarative Services annotations or (probably more convenient in this case) by registering the wrapper services using `BundleContext.registerService.`

I'm happy to create an example of that, in this module's test code for example, to show how it can work. I think you'll find that it's the same as implementing your scalars provider, except that the provider registers OSGi services instead of making the converters available via an API that we can avoid. I'd say we get the same result but with less code and a smaller API surface.

> IMHO scalars should not be defined globally, but per endpoint. Typically, they are part of the schema...

I agree that the set of active scalars is often schema-dependent. We might use a `@namespace(scalars="mynamespace")` directive or something similar to namespace them. I think that's an orthogonal concern to what we are discussing here.

**Comment by bdelacretaz** [comment]
Let me know if the previous comment makes sense in your context, if you need me to provide an example of the wrappers that I mentioned, or if smaller tweaks to the API would make your life easier!

**Comment by stefangrimm** [comment]
I will see if I can come up with something along the existing implementation and check if it really works (biggest problem I see is that we can't register the services "early enough", as the scalars will only be available when the schema is created - which potentially could be too late, but let's see if I'm just overly panicky here).

If you have pointers to some existing code that does similar things, feel free to share. No need to create something new now. Wanted to improve my OSGi capabilities anyway ...

**Comment by bdelacretaz** [comment]
I have commented in [SLING-9550](https://issues.apache.org/jira/browse/SLING-9550?focusedCommentId=17150946&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17150946) with a proposal for namespacing the scalars.

**Comment by bdelacretaz** [comment]
I have merged the feature branch into master and marked https://issues.apache.org/jira/browse/SLING-9550 resolved accordingly, for basic scalars support. We can still rediscuss the API if needed, preferably on the Sling dev list as others might also have an opinion.

## How to register external converters without the SlingScalarConvertersProvider

As for registering wrappers for existing scalar converters (to avoid the `SlingScalarConvertersProvider` service that this PR suggests) it's basically what the [TestUtil.registerSlingScalarConverter](https://github.com/apache/sling-org-apache-sling-graphql-core/blob/9bc163cd7eab8b7c3a1ab107656253aa9ebb5e2a/src/test/java/org/apache/sling/graphql/core/mocks/TestUtil.java#L39) method does:

    BundleContext bc = ...injected in activate method
    SlingScalarConverter converter = new ConverterWrapper(converterToWrap);
    final Dictionary<String, Object> props = new Hashtable<>();
    props.put(SlingScalarConverter.NAME_SERVICE_PROPERTY, name);
    bc.registerService(SlingScalarConverter.class, converter, props);

So if you have a set of `converterToWrap`, just run this in a loop. This can happen in the `activate` method of an OSGi Component for example, and the converters just need to be registered before running the first GraphQL query.

**Comment by bdelacretaz** [comment]
I'm closing this due to no activity and as https://issues.apache.org/jira/browse/SLING-9550 closed now. We can reopen or create a new one if needed.

**Pull request closed without merging**